File: implementations/httpcat-node/src/logic.ts
import asciify from "asciify-image";
import { getReasonPhrase } from "http-status-codes";

export enum PetType {
	Cat = "cat",
	Dog = "dog",
}

const HTTP_CAT_ROOT_URL = "https://http.cat";
const HTTP_DOG_ROOT_URL = "https://http.dog";

export function getPetCodeImageUrl(
	petType: PetType,
	statusCode: number,
): string {
	if (petType === PetType.Cat) {
		return `${HTTP_CAT_ROOT_URL}/images/${statusCode}.jpg`;
	}

	return `${HTTP_DOG_ROOT_URL}/${statusCode}.jpg`;
}

export async function fetchPetCodeImage(
	petType: PetType,
	statusCode: number,
): Promise<ArrayBuffer> {
	const responseData = await fetch(getPetCodeImageUrl(petType, statusCode));
	return responseData.arrayBuffer();
}

export async function convertImageToAscii(
	imageData: ArrayBuffer,
	width: number,
): Promise<string> {
	const data = await asciify(Buffer.from(imageData), {
		color: true,
		fit: "width",
		width,
	});

	// Replace ";" with a dot '.' to "improve" rendering.
	return (data as string).replace(/;/g, ".");
}

export function isStatusCodeValid(statusCode: number): boolean {
	// Not pretty, I know.
	try {
		getReasonPhrase(statusCode);
		return true;
	} catch (_e) {
		return false;
	}
}

File: implementations/httpcat-dart/lib/logic.dart
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:enough_ascii_art/enough_ascii_art.dart' as art;
import 'package:image/image.dart' as image;
import 'package:http_status_code/http_status_code.dart' as hsc;

enum PetType { cat, dog }

const httpCatRootUrl = "https://http.cat";
const httpDogRootUrl = "https://http.dog";

bool isStatusCodeValid(int statusCode) {
  return hsc.getStatusMessage(statusCode) != "Unknown Status Code";
}

Uri getPetCodeImageUrl(PetType petType, int statusCode) {
  if (petType == PetType.cat) {
    return Uri.parse("$httpCatRootUrl/images/$statusCode.jpg");
  }
  return Uri.parse("$httpDogRootUrl/$statusCode.jpg");
}

Future<Uint8List> fetchPetCodeImage(PetType petType, int statusCode) async {
  final data = await http.get(getPetCodeImageUrl(petType, statusCode));
  return data.bodyBytes;
}

String convertImageToAscii(Uint8List imageData, int width) {
  final img = image.decodeImage(imageData)!;
  var ascii = art.convertImage(img, maxWidth: width, invert: true);
  return ascii;
}
